drivers/sensor/bmm150/bmm150.c

```diff
@@ -0,0 +1,602 @@
+		BMM150_SHIFT_Z_L;
+
+	rhall = sys_le16_to_cpu(values[BMM150_RHALL]) >>
+		BMM150_SHIFT_RHALL_L;
+
+	drv_data->sample_x = raw_x;
+	drv_data->sample_y = raw_y;
+	drv_data->sample_z = raw_z;
+	drv_data->rhall = rhall;
+
+	return 0;
+}
+
+static void bmm150_convert(struct sensor_value *val, int raw_val)
+{
+	/* val = raw_val /1600 */
+	val->val1 = raw_val / 1600;
+	val->val2 = ((s32_t)raw_val * (1000000 / 1600)) % 1000000;
+}
+
+static int bmm150_channel_get(struct device *dev, 
+				enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct bmm150_data *drv_data = dev->driver_data;
+        
+	switch (chan) {
+	case SENSOR_CHAN_MAGN_X:
+		drv_data->sample_x = bmm150_compensate_xy(&drv_data->tregs, 
+				drv_data->sample_x,drv_data->rhall,
+                                                  true);
+		bmm150_convert(val, drv_data->sample_x);
+		break;
+	case SENSOR_CHAN_MAGN_Y:
+		drv_data->sample_y = bmm150_compensate_xy(&drv_data->tregs, 
+				drv_data->sample_y,drv_data->rhall,
+                                                  false);
+		bmm150_convert(val, drv_data->sample_y);
+		break;
+	case SENSOR_CHAN_MAGN_Z:
+		drv_data->sample_x = bmm150_compensate_z(&drv_data->tregs, 
+				drv_data->sample_z,drv_data->rhall);
+		bmm150_convert(val, drv_data->sample_x);
+		break;
+	case SENSOR_CHAN_MAGN_ANY:
+		drv_data->sample_x = bmm150_compensate_xy(&drv_data->tregs, 
+				drv_data->sample_x,drv_data->rhall,
+                                                  true);
+		drv_data->sample_y = bmm150_compensate_xy(&drv_data->tregs, 
+				drv_data->sample_y,drv_data->rhall,
+                                                  false);
+		drv_data->sample_x = bmm150_compensate_z(&drv_data->tregs, 
+				drv_data->sample_z,drv_data->rhall);
+		bmm150_convert(val, drv_data->sample_x);
+		bmm150_convert(val + 1, drv_data->sample_y);
+		bmm150_convert(val + 2, drv_data->sample_z);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+#if defined(BMM150_SET_ATTR_REP)
+static inline int bmm150_attr_set_rep(struct device *dev,
+                                           enum sensor_channel chan,
+                                           const struct sensor_value *val)
+{
+        struct bmm150_data *data = dev->driver_data;
+        int max_odr;
+
+        switch (chan) {
+#if defined(CONFIG_BMM150_SAMPLING_REP_XY)
+        case SENSOR_CHAN_MAGN_X:
+        case SENSOR_CHAN_MAGN_Y:
+                if (val->val1 < 1 || val->val1 > BMM150_REP_XY_MAX) {
+                        return -EINVAL;
+                }
+
+                if (bmm150_compute_max_odr(dev, val->val1, 0,
+                                                &max_odr) < 0) {
+                        return -EIO;
+                }
+
+                if (data->odr <= 0) {
+                        if (bmm150_read_odr(dev) < 0) {
+                                return -EIO;
+                        }
+                }
+
+                if (data->odr > max_odr) {
```

Comment:
can max_odr be below 0? 
If not, you may put this test as an else if () to previous if()